Lots of changes to the Tile Pipeline

Bdtrotte opened the pull request:
Mainly: Now TileTemplate emits a signal to tiles when changed,
whenever a Tile changed, it emits a signal to the TileMap,
which forwards it to whoever wants it (mapview/meshview...)
As well, the TileMap's array of tiles now stay the same, and the tiles
are just changed, which previously was not the case.

Bdtrotte committed:
Lots of changes to the Tile Pipeline
Mainly: Now TileTemplate emits a signal to tiles when changed,
whenever a Tile changed, it emits a signal to the TileMap,
which forwards it to whoever wants it (mapview/meshview...)
As well, the TileMap's array of tiles now stay the same, and the tiles
are just changed, which previously was not the case.

timoffex commented on WallsAndHoles/tile.cpp:
```diff
@@ -10,7 +10,20 @@ Tile::Tile(const SharedTileTemplate tileTemplate,
+
+Tile::~Tile()
+{
+    mTileTemplate->disconnect(this);
```
"A signal-slot connection is removed when either of the objects involved are destroyed."  http://doc.qt.io/qt-5/qobject.html#disconnect-2

So this line is not necessary.

timoffex commented on WallsAndHoles/tilemap.cpp:
```diff
@@ -1,19 +1,17 @@
-    mMap = (QSharedPointer<Tile> **)malloc(sizeof(QSharedPointer<Tile> *) * mapSize.width());
-    for (int x = 0; x < mapSize.width(); ++x) {
-        mMap[x] = (QSharedPointer<Tile> *)malloc(sizeof(QSharedPointer<Tile>) * mapSize.height());
+    mMap = (Tile ***)malloc(sizeof(Tile **) * mMapSize.width());
```
malloc() and free() are C-style coding. I recommend using

mMap = new Tile**[ mMapSize.width() ];
mMap[x] = new Tile*[ mMapSize.height() ];  for all x
mMap[x][y] = new Tile(...);  for all x, y

and then don't forget to add to your destructor:

delete mMap[x][y]; for all x,y
delete [] mMap[x]; for all x
delete [] mMap;


Even better: make a class for 2D arrays that uses vectors to implement them. This moves all the messy vector<vector<>> stuff outside the Tile code, allows easy resizability, and gets rid of any chance of a memory leak (as long as you use a vector<vector<shared pointer>>)

timoffex replied:
right now your destructor does not free mMap completely

timoffex committed:
Added Array2D

In TileMap, change the type of mMap to

Array2D<QSharedPointer<Tile>>

In the constructor for TileMap, initialize mMap with
  mMap(width, height)

Then just loop over all possible (x,y) pairs and type mMap(x, y) =
QSharedPointer<Tile>::create(…);

(yes, use parentheses for access!)

Bdtrotte committed:
Using Array2D<QSharedPointer<Tile>> for TileMap
